lang/src/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt

```diff
@@ -524,6 +525,93 @@ internal class StaticTypeInferenceVisitorTransform(
+                                                lastType.rangeConstraint == IntType.IntRangeConstraint.UNCONSTRAINED -> lastType
+                                                currentType.rangeConstraint == IntType.IntRangeConstraint.UNCONSTRAINED -> currentType
+                                                lastType.rangeConstraint.numBytes > currentType.rangeConstraint.numBytes -> lastType
+                                                else -> currentType
+                                            }
+                                        }
+                                        is FloatType -> StaticType.FLOAT
+                                        is DecimalType -> StaticType.DECIMAL // TODO:  account for decimal precision
+                                        // only missing and null are possible because of argTypeCheck()
+                                        else -> currentType
+                                    }
+                                }
+                                is FloatType -> {
+                                    when (lastType) {
+                                        is IntType -> StaticType.FLOAT
+                                        is FloatType -> StaticType.FLOAT
+                                        is DecimalType -> StaticType.DECIMAL // TODO:  account for decimal precision
+                                        is NullType, is MissingType -> currentType
+                                        else -> error("Unsupported argument type for Aggregate Function")
+                                    }
+                                }
+                                is DecimalType -> {
+                                    when (lastType) {
+                                        is IntType -> StaticType.DECIMAL // TODO:  account for decimal precision
+                                        is FloatType -> StaticType.DECIMAL // TODO:  account for decimal precision
+                                        is DecimalType -> StaticType.DECIMAL // TODO:  account for decimal precision
+                                        is NullType, is MissingType -> currentType
+                                        else -> error("Unsupported argument type for Aggregate Function")
+                                    }
+                                }
+                                else ->
+                                    // this should not be reached, only exists to segment the logic.
+                                    error(
+                                        "Internal Error: SUM function only support Number Type. " +
+                                            "This probably indicates an bug in type inferencer."
+                                    )
+                            }
+                        }
+                    } else {
+                        error("Aggregate function SUM has incompatible input type")
+                    }
+                }
```

Comment:
For the data type 'error' conditions in the static type inferencer (e.g. when a non-numeric type is passed to an aggregation function that expects a numeric type), we've previously added the error to the `ProblemHandler` and defined a continuation behavior. This allows us to find other inferencing errors in the PartiQL statement rather than failing fast on the first data type mismatch.

For example, in an expression such as `'foo' + 'bar' + 'baz'`, the inferencer would identify a data type mismatch at both of the `+` operations since it expects a numeric type. We defined a continuation behavior for the `+` (and other arithmetic operations) to return the union of numeric types ([source](https://github.com/partiql/partiql-lang-kotlin/blob/d03098b5dfd86e69cdc2faf05ba0da96b61aa9d3/lang/src/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt#L701-L702)).

I think a similar logic should apply to this case for `SUM`.